pkg/apis/eventing/v1alpha1/provisioner_types.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"encoding/json"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Provisioner provisions.
+type Provisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ProvisionerSpec   `json:"spec"`
+	Status ProvisionerStatus `json:"status"`
+}
+
+// Check that Provisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*Provisioner)(nil)
+var _ apis.Defaultable = (*Provisioner)(nil)
+var _ runtime.Object = (*Provisioner)(nil)
+var _ webhook.GenericCRD = (*Provisioner)(nil)
+
+// ProvisionerSpec is the spec for a Provisioner resource.
+type ProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Type is the type of the resource to be provisioned.
+	// +required
+	Type runtime.TypeMeta `json:"type"`
+}
+
+type ProvisionerConditionType string
+
+const (
+	// ProvisionerConditionReady specifies that the resource is ready.
+	ProvisionerConditionReady ProvisionerConditionType = "Ready"
+)
+
+// ProvisionerConditionStatus describes the state of this resource at a point in time.
+type ProvisionerConditionStatus struct {
+	// Type of condition.
+	// +required
+	Type ProvisionerConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status corev1.ConditionStatus `json:"status"`
+
+	// The last time this condition was updated.
+	// +optional
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
```

Comment:
Do we need this? I understand why we want `LastUpdateTime` to tell if the controller is alive.

Reply:
gone.